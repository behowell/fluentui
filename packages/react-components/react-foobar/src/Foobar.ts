export * from './components/Foobar/index';
