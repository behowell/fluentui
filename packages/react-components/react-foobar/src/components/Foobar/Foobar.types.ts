import type { ComponentProps, ComponentState, Slot } from '@fluentui/react-utilities';

export type FoobarSlots = {
  root: Slot<'div'>;
};

/**
 * Foobar Props
 */
export type FoobarProps = ComponentProps<FoobarSlots> & {};

/**
 * State used in rendering Foobar
 */
export type FoobarState = ComponentState<FoobarSlots>;
// TODO: Remove semicolon from previous line, uncomment next line, and provide union of props to pick from FoobarProps.
// & Required<Pick<FoobarProps, 'propName'>>
