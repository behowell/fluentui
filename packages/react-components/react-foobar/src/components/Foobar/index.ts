export * from './Foobar';
export * from './Foobar.types';
export * from './renderFoobar';
export * from './useFoobar';
export * from './useFoobarStyles';
