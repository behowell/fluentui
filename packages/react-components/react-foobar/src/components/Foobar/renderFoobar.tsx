import * as React from 'react';
import { getSlots } from '@fluentui/react-utilities';
import type { FoobarState, FoobarSlots } from './Foobar.types';

/**
 * Render the final JSX of Foobar
 */
export const renderFoobar_unstable = (state: FoobarState) => {
  const { slots, slotProps } = getSlots<FoobarSlots>(state);

  // TODO Add additional slots in the appropriate place
  return <slots.root {...slotProps.root} />;
};
