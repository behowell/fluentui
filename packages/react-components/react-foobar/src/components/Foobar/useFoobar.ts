import * as React from 'react';
import { getNativeElementProps } from '@fluentui/react-utilities';
import type { FoobarProps, FoobarState } from './Foobar.types';

/**
 * Create the state required to render Foobar.
 *
 * The returned state can be modified with hooks such as useFoobarStyles_unstable,
 * before being passed to renderFoobar_unstable.
 *
 * @param props - props from this instance of Foobar
 * @param ref - reference to root HTMLElement of Foobar
 */
export const useFoobar_unstable = (props: FoobarProps, ref: React.Ref<HTMLElement>): FoobarState => {
  return {
    // TODO add appropriate props/defaults
    components: {
      // TODO add each slot's element type or component
      root: 'div',
    },
    // TODO add appropriate slots, for example:
    // mySlot: resolveShorthand(props.mySlot),
    root: getNativeElementProps('div', {
      ref,
      ...props,
    }),
  };
};
