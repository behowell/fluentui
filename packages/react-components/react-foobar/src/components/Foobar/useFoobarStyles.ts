import { makeStyles, mergeClasses } from '@griffel/react';
import type { FoobarSlots, FoobarState } from './Foobar.types';
import type { SlotClassNames } from '@fluentui/react-utilities';

export const foobarClassNames: SlotClassNames<FoobarSlots> = {
  root: 'fui-Foobar',
  // TODO: add class names for all slots on FoobarSlots.
  // Should be of the form `<slotName>: 'fui-Foobar__<slotName>`
};

/**
 * Styles for the root slot
 */
const useStyles = makeStyles({
  root: {
    // TODO Add default styles for the root element
  },

  // TODO add additional classes for different states and/or slots
});

/**
 * Apply styling to the Foobar slots based on the state
 */
export const useFoobarStyles_unstable = (state: FoobarState): FoobarState => {
  const styles = useStyles();
  state.root.className = mergeClasses(foobarClassNames.root, styles.root, state.root.className);

  // TODO Add class names to slots, for example:
  // state.mySlot.className = mergeClasses(styles.mySlot, state.mySlot.className);

  return state;
};
