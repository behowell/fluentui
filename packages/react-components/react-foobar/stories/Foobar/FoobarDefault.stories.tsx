import * as React from 'react';
import { Foobar, FoobarProps } from '@fluentui/react-foobar';

export const Default = (props: Partial<FoobarProps>) => <Foobar {...props} />;
