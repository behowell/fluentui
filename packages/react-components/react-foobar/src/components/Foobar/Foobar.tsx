import * as React from 'react';
import { useFoobar_unstable } from './useFoobar';
import { renderFoobar_unstable } from './renderFoobar';
import { useFoobarStyles_unstable } from './useFoobarStyles';
import type { FoobarProps } from './Foobar.types';
import type { ForwardRefComponent } from '@fluentui/react-utilities';

/**
 * Foobar component - TODO: add more docs
 */
export const Foobar: ForwardRefComponent<FoobarProps> = React.forwardRef((props, ref) => {
  const state = useFoobar_unstable(props, ref);

  useFoobarStyles_unstable(state);
  return renderFoobar_unstable(state);
});

Foobar.displayName = 'Foobar';
