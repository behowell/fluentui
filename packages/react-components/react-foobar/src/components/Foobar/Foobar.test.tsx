import * as React from 'react';
import { render } from '@testing-library/react';
import { Foobar } from './Foobar';
import { isConformant } from '../../testing/isConformant';

describe('Foobar', () => {
  isConformant({
    Component: Foobar,
    displayName: 'Foobar',
  });

  // TODO add more tests here, and create visual regression tests in /apps/vr-tests

  it('renders a default state', () => {
    const result = render(<Foobar>Default Foobar</Foobar>);
    expect(result.container).toMatchSnapshot();
  });
});
