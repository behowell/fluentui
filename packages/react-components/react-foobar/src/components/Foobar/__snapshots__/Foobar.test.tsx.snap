// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Foobar renders a default state 1`] = `
<div>
  <div
    class="fui-Foobar"
  >
    Default Foobar
  </div>
</div>
`;
